## 4.7 Deriving New Variables

### Contents

### 1. Exercise 4.7 procedure
### - Importing libraries and files
### - If-statement
### - If-statement with loc()
### - If-statement with for-loops

### 2. Exercise 4.7 task

#### 1. If you haven’t done so already, complete the instructions in the Exercise for creating the “price_label” and “busiest_day” columns.

#### Importing libraries and files

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\boss4048\Documents\10-2024 Instacart Basket Analysis'

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared data', 'ords_prods_merge.pkl'))

In [4]:
# Limit the data from row 0 to row 1,000,000
df = ords_prods_merge[:1000000]

In [5]:
df

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0                  0             0   2539329        1             1   
1                  0             0   2539329        1             1   
2                  0             0   2539329        1             1   
3                  0             0   2539329        1             1   
4                  0             0   2539329        1             1   
...              ...           ...       ...      ...           ...   
999995        106983        106983   2789027     6474             2   
999996        106983        106983   2789027     6474             2   
999997        106983        106983   2789027     6474             2   
999998        106983        106983   2789027     6474             2   
999999        106983        106983   2789027     6474             2   

        orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                        2                  8                          NaN   
1                        2                  8                          NaN   
2                        2                  8                          NaN   
3                        2                  8                          NaN   
4                        2                  8                          NaN   
...                    ...                ...                          ...   
999995                   3                  8                         30.0   
999996                   3                  8                         30.0   
999997                   3                  8                         30.0   
999998                   3                  8                         30.0   
999999                   3                  8                         30.0   

        product_id  add_to_cart_order  reordered  Unnamed: 0_y  \
0              196                  1          0           195   
1            14084                  2          0         14084   
2            12427                  3          0         12427   
3            26088                  4          0         26089   
4            26405                  5          0         26406   
...            ...                ...        ...           ...   
999995        3020                  7          0          3020   
999996       35823                  8          0         35827   
999997       27681                  9          0         27683   
999998        2091                 10          0          2091   
999999       44199                 11          0         44203   

                                             product_name  aisle_id  \
0                                                    Soda        77   
1                 Organic Unsweetened Vanilla Almond Milk        91   
2                                     Original Beef Jerky        23   
3                              Aged White Cheddar Popcorn        23   
4                        XL Pick-A-Size Paper Towel Rolls        54   
...                                                   ...       ...   
999995                 Broccoli & Apple Stage 2 Baby Food        92   
999996  Complete Foaming Antibacterial Hand Soap Varie...       127   
999997               Lower Sugar Instant Oatmeal  Variety       130   
999998                   Almond Nut & Rice Cracker Snacks        78   
999999        Toll House Mini Chocolate Chip Cookie Dough       105   

        department_id  prices _merge  
0                   7     9.0   both  
1                  16    12.5   both  
2                  19     4.4   both  
3                  19     4.7   both  
4                  17     1.0   both  
...               ...     ...    ...  
999995             18    12.9   both  
999996             11     7.3   both  
999997             14     3.5   both  
999998             19     6.8   both  
999999             13     2.9   both  

[1000000 rows x 17 columns]

#### If else-statement

In [6]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
## Applying if-statement into the function we created 
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\boss4048\AppData\Local\Temp\ipykernel_25588\958781529.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [8]:
df.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             0             0   2539329        1             1   
2             0             0   2539329        1             1   
3             0             0   2539329        1             1   
4             0             0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                   2                  8                          NaN   
1                   2                  8                          NaN   
2                   2                  8                          NaN   
3                   2                  8                          NaN   
4                   2                  8                          NaN   

   product_id  add_to_cart_order  reordered  Unnamed: 0_y  \
0         196                  1          0           195   
1       14084                  2          0         14084   
2       12427                  3          0         12427   
3       26088                  4          0         26089   
4       26405                  5          0         26406   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge        price_range  
0   both  Mid-range product  
1   both  Mid-range product  
2   both  Low-range product  
3   both  Low-range product  
4   both  Low-range product

In [9]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673183
Low-range product    314392
High range            12425
Name: count, dtype: int64

In [10]:
df['prices'].max()

99999.0

#### If-statement with loc function

In [11]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [12]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [13]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [14]:
# Check the counts of each value after applying if-statement with loc function
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

#### If-statements with For-Loops

In [15]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [16]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")


In [17]:
# The result is quite long
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [18]:
# Create a new column within ords_prods_merge dataframe and set it equal to result
ords_prods_merge['busiest_day'] = result

In [19]:
# Check counts of each value of our flag column ('busiest_day')
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

In [20]:
# Cross-check with count of 'orders_day_of_week' column
ords_prods_merge['orders_day_of_week'].count()

32404859

The sum counts of the flag column (busiest_day).
Regularly busy 22,416,875 + Busiest day 6,204,182 + Least busy 3,783,802 = 32,404,859) is equal to the sum count of the column we created from (orders_day_of_week).

#### 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [21]:
# Create a new list to save the result after running the for-loop according to the task's assignment
result2 = []
for value in ords_prods_merge['orders_day_of_week']:
  if value in (0,1):
    result2.append('Busiest days')
  elif value in (3,4):
    result2.append('Slowest days')
  else:
    result2.append('Regularly busy')

Explaination: 
I decided to use if-statement with for loop. From the counts of each value from 'orders_day_of_week' column as seen above, 
Saturday (0) and Sunday (1) are the busiest days.
Tuesday (3) and Wednesday (4) are the slowest days.
Other days (2,5,6) are Regularly busy days.

In [22]:
# Result variable we created is a list. we need to save it into a dataframe in order to be able to execute any commands in Python
# Save the list into ords_prods_merge dataframe. By coding below, we are creating a new column ('Busiest days')
ords_prods_merge['Busiest days'] = result2

In [23]:
# Use .head() function to check the result. We can see the last column ('Busiest days') was created
ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             0             0   2539329        1             1   
2             0             0   2539329        1             1   
3             0             0   2539329        1             1   
4             0             0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                   2                  8                          NaN   
1                   2                  8                          NaN   
2                   2                  8                          NaN   
3                   2                  8                          NaN   
4                   2                  8                          NaN   

   product_id  add_to_cart_order  reordered  Unnamed: 0_y  \
0         196                  1          0           195   
1       14084                  2          0         14084   
2       12427                  3          0         12427   
3       26088                  4          0         26089   
4       26405                  5          0         26406   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge    price_range_loc     busiest_day    Busiest days  
0   both  Mid-range product  Regularly busy  Regularly busy  
1   both  Mid-range product  Regularly busy  Regularly busy  
2   both  Low-range product  Regularly busy  Regularly busy  
3   both  Low-range product  Regularly busy  Regularly busy  
4   both  Low-range product  Regularly busy  Regularly busy

#### 3. Check the values of this new column for accuracy. Note any observations in markdown format.

In [24]:
ords_prods_merge['Busiest days'].value_counts(dropna = False)

Busiest days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [25]:
ords_prods_merge['Busiest days'].count()

32404859

Observation
- Sum of each value from the new column (by using .value_counts()) match the total count of the whole new column (by using .count())
Regularly busy (12,916,111) + Busiest days (11,864,412) + Slowest days (7,624,336) = Total count (32,404,859)

This is proving that we are not missing any values after executing the code (If-statements with For-Loops).

- The new column (Busiest days) was created by grouping the values from the previous column (busiest_day), so the sum values should match.

Regularly busy (12,916,111) = 6 (4,496,490) + 2 (4,213,830) + 5 (4,205,791)

Busiest days (11,864,412) = 0 (6,204,182) + 1 (5,660,230)

Slowest days (7,624,336) = 3 (3,840,534) + 4 (3,783,802)

All values in each group match!

#### 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [26]:
# Used .value.counts() of 'order_hour_of_day' column to see frequency of each hour 
ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

Split one third of values from 'order_hour_of_day' column into each group.

Hour: 10, 11, 14, 15, 13, 12, 16, 9 = Most orders

Hour: 17, 8, 18, 19, 20, 7, 21, 22  = Average orders

Hour: 23, 6, 0, 1, 5, 2, 4, 3       = Fewest orders

In [27]:
# Create a new list to save the result after running the for-loop according to the task's assignment
result3 = []
for value in ords_prods_merge['order_hour_of_day']:
  if value in (10, 11, 14, 15, 13, 12, 16, 9):
    result3.append('Most orders')
  elif value in (17, 8, 18, 19, 20, 7, 21, 22):
    result3.append('Average orders')
  else:
    result3.append('Fewest orders')

In [28]:
# Create a new column containing the labels called 'busiest_period_of_day.'
ords_prods_merge['busiest_period_of_day'] = result3

In [29]:
# Check the result if the new column ('busiest_period_of_day') is created
ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id  order_number  \
0             0             0   2539329        1             1   
1             0             0   2539329        1             1   
2             0             0   2539329        1             1   
3             0             0   2539329        1             1   
4             0             0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  total_days_since_last_order  \
0                   2                  8                          NaN   
1                   2                  8                          NaN   
2                   2                  8                          NaN   
3                   2                  8                          NaN   
4                   2                  8                          NaN   

   product_id  add_to_cart_order  ...  Unnamed: 0_y  \
0         196                  1  ...           195   
1       14084                  2  ...         14084   
2       12427                  3  ...         12427   
3       26088                  4  ...         26089   
4       26405                  5  ...         26406   

                              product_name aisle_id  department_id  prices  \
0                                     Soda       77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk       91             16    12.5   
2                      Original Beef Jerky       23             19     4.4   
3               Aged White Cheddar Popcorn       23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls       54             17     1.0   

   _merge    price_range_loc     busiest_day    Busiest days  \
0    both  Mid-range product  Regularly busy  Regularly busy   
1    both  Mid-range product  Regularly busy  Regularly busy   
2    both  Low-range product  Regularly busy  Regularly busy   
3    both  Low-range product  Regularly busy  Regularly busy   
4    both  Low-range product  Regularly busy  Regularly busy   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  

[5 rows x 21 columns]

#### 5. Print the frequency for this new column.

In [30]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

In [31]:
21118071 + 9997651 + 1289137

32404859

The sum of all values matches the total count we calculated earlier.

#### 6. Ensure your notebook is clean and structured and that your code is well commented.

The notebook looks clean and structured with comments and markdown explaination.

#### 7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [32]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_derived.pkl'))